File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline

N, M = map(int, input().split())

# 인접 리스트로 그래프를 저장할 2차원 리스트
A = [[] for _ in range(N + 1)]

# 방문 여부를 체크하는 리스트
# 모든 노드를 방문하지 않은 상태(False)로 초기화
visited = [False] * (N + 1)

def DFS(v):
    visited[v] = True # 방문한 노드를 True로 표시
    # 현재 v노드에 연결된 모든 노드를 탐색 
    for i in A[v]:
        # e.g. v = 1, A[1] = [2, 5]라면
        # 1번 노드에서 인접한 2번 5번 노드에 대해 재귀함수 호출
        if not visited[i]:
            DFS(i)

# 간선의 개수만큼 입력받음
for _ in range(M):
    # 양 끝점 입력
    s, e = map(int, input().split())

    # 무방향 그래프이므로 서로 간선을 추가함
    A[s].append(e)
    A[e].append(s)

# 연결 요소를 카운트할 변수
count = 0

# 모든 정점의 개수인 N만큼 반복
for i in range(1, N + 1):
    # 인접 리스트 A에서 방문하지 않은 노드가 있다면
    if not visited[i]:
        count += 1 # 연결 요소 카운트
        DFS(i) # DFS 함수를 통해 연결된 모든 노드 방문 처리

print(count)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<Integer>[] A;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        A = new ArrayList[N + 1]; // 인접 리스트 배열 인스턴스 생성
        visited = new boolean[N + 1]; // 방문 리스트 생성 (자동으로 false 초기화)

        // 배열의 각 요소에 대해 ArrayList<Integer> 객체로 인스턴스 생성 -> 인접 리스트 초기화
        for (int i = 1; i <= N; i++) {
            A[i] = new ArrayList<Integer>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            // 양방향 에지이므로 양쪽에 간선 추가
            A[start].add(end);
            A[end].add(start);
        }

        int count = 0;

        for (int i = 1; i <= N; i++) {
            if (!visited[i]) {
                count++;
                DFS(i);
            }
        }

        System.out.println(count);
    }

    static void DFS(int v) {
        visited[v] = true;
        for (int i : A[v]) {
            if (!visited[i]) {
                DFS(i);
            }
        }
    }
}
